openwisp_controller/config/api/serializers.py

```diff
@@ -0,0 +1,217 @@
+
+from django.db import transaction
+from django.utils.translation import ugettext_lazy as _
+from rest_framework import serializers
+from swapper import load_model
+
+from openwisp_users.api.mixins import FilterSerializerByOrgManaged
+from openwisp_utils.api.serializers import ValidatedModelSerializer
+
+Template = load_model('config', 'Template')
+Vpn = load_model('config', 'Vpn')
+Device = load_model('config', 'Device')
+Config = load_model('config', 'Config')
+TemplateTag = load_model('config', 'TemplateTag')
+Organization = load_model('openwisp_users', 'Organization')
+
+
+class BaseMeta:
+    read_only_fields = ['created', 'modified']
+
+
+class BaseSerializer(FilterSerializerByOrgManaged, ValidatedModelSerializer):
+    pass
+
+
+class TemplateSerializer(BaseSerializer):
+    config = serializers.JSONField(initial={})
+    tags = serializers.StringRelatedField(many=True, read_only=True)
+    default_values = serializers.JSONField(required=False, initial={})
+
+    class Meta(BaseMeta):
+        model = Template
+        fields = [
+            'id',
+            'name',
+            'tags',
+            'organization',
+            'type',
+            'backend',
+            'default',
+            'required',
+            'default_values',
+            'config',
+            'created',
+            'modified',
+        ]
+
+
+class VpnSerializer(BaseSerializer):
+    config = serializers.JSONField(initial={})
+
+    class Meta(BaseMeta):
+        model = Vpn
+        fields = [
+            'id',
+            'name',
+            'host',
+            'organization',
+            'key',
+            'ca',
+            'cert',
+            'backend',
+            'notes',
+            'dh',
+            'config',
+        ]
+
+
+class BaseJsonField(serializers.Field):
+    def to_internal_value(self, data):
+        if type(data) is str:
+            if len(data) < 2:
+                data = '{}'
+            return ast.literal_eval(data)
+        else:
+            return data
+
+
+class JsonConfigField(BaseJsonField):
+    def to_representation(self, value):
+        return json.loads(json.dumps(value))
+
+
+class JsonContextField(BaseJsonField):
+    def to_representation(self, value):
+        return dict(value)
+
+
+class DeviceConfigSerializer(serializers.ModelSerializer):
+    config = JsonConfigField(
+        help_text='Configuration in NetJSON format',
+        style={'base_template': 'textarea.html', 'placeholder': '{}'},
+    )
+    context = JsonContextField(
+        help_text='Configuration in NetJSON format',
+        style={'base_template': 'textarea.html', 'placeholder': '{}'},
+    )
+
+    class Meta:
+        model = Config
+        fields = ['backend', 'status', 'templates', 'context', 'config']
+        extra_kwargs = {'status': {'read_only': True}}
+
+
+class DeviceListSerializer(FilterSerializerByOrgManaged, serializers.ModelSerializer):
+    config = DeviceConfigSerializer(write_only=True, required=False)
```

Comment:
you must find a way to call the full_clean() method of the models of this serializer because there's a lot of model validation done in `Device` and `Config`.

For example, if you try to create a device with name `T E S T`, the API will let you create it, but the admin doesn't, because it's not a valid hostname. Please add a basic test for this.

Reply:
Added a test for this.

Reply:
@ManishShah120 did you see one of my previous reviews I mentioned the following?

> It may be important to show also config.status and config.backend in the list, since these two pieces of information are quite important. Can you add a limited config serializer in the list serializer too @ManishShah120? (Keep in mind to use select_related('config') to keep the number of generated queries under control).

Did you see this?

Reply:
Yes it has been implemented.